### PR TITLE
Dynamic instead of hard-coded port numbers

### PR DESCRIPTION
**Pull request by loopmode**
This change adds a new dependency, https://www.npmjs.com/package/get-port, and uses it to dynamically retrieve a free port number for webpack-dev-server and the chrome inspector.

It enables us to work on multiple `electron-webpack` projects simultanously - currently not possible due to EADDRINUSE errors.

My commit message states that the change resolves issue https://github.com/electron-userland/electron-webpack/issues/25. That is not necessarily true.
We still cannot specify a particular port number, but at least it doesn't collide anymore.

It might be an issue for users who need to know the exact used port, or who must specify it for some reason.
That, however, is a separate feature and should be added in the future, for example via CLI args.

### CODE REVIEW

File: packages/electron-webpack/src/dev/dev-runner.ts [hunk 1]
```diff
@@ -136,9 +137,10 @@ main()
     console.error(error)
   })
 
-function startElectron() {
+async function startElectron() {
   const electronArgs = process.env.ELECTRON_ARGS
-  const args = electronArgs != null && electronArgs.length > 0 ? JSON.parse(electronArgs) : ["--inspect=5858"]
+  const port = await getPort();
```
Comment (loopmode) [inline packages/electron-webpack/src/dev/dev-runner.ts]: Please note that while `get-port` supports a "favorite" port number via `getPort({port: 5858})`, the typescript definitions from `@types/get-port` do not support it yet.
So, in order to make compilation possible, we currently do not specify a port and just use whatever comes back instead.

**Comment by loopmode** [comment]
Ugh. Of course we need to know the port ourselves when we do this in the main process:
```
const mainWindow = new BrowserWindow();
let url = isDevelopment
    ? 'http://localhost:9080' // <-- Code won't work in case of value other than 9080
    : `file://${__dirname}/index.html`;
win.loadURL(url);
```
Need to figure out how to do that. Closing PR.

**Comment by loopmode** [comment]
@develar We spoke about this. I agreed with you that no user actually cares about such an internal detail, about the actual port number. But as you see.. it actually is relevant technically :)

Not sure how leight-weight and easy a bridge between the user's main process file and the dynamically retrieved port number would be. Will think about it.

On the other hand, it would be a trivial change to pick it up from an `.env` file or the CLI args and it would be entirely declarative and deterministic.

**Comment by SimulatedGREG** [comment]
Would probably be easier just to create an option for it in the custom
configurations.

**Comment by loopmode** [comment]
Not sure how that works. Would the user be able to specify that in his project, outside of the actual `electron-webpack` package?

**Comment by SimulatedGREG** [comment]
https://webpack.electron.build/configuration

Just like the other configurations that are available.

**Comment by loopmode** [comment]
Ah right, perfect. I think that's the way to go then.

**Comment by develar** [comment]
Main process has IPC channel connection to dev runner. Pass WDS port in the same way — using some internal env name (e.g. we currently pass `ELECTRON_HMR_SOCKET_PATH`).

**Comment by develar** [comment]
No need to close PR — just add commits. The only required change — I suggest to introduce internal env `ELECTRON_WDS_PORT` and use it in the main code instead of `9080`.

**Comment by loopmode** [comment]
The used port is now exposed via `process.env.ELECTRON_WDS_PORT` as suggested by @develar .

Regarding https://github.com/electron-userland/electron-webpack/issues/25:
With this change, what we _need_ to do in our projects is this:
```
    let url = isDevelopment ? `http://localhost:${process.env.ELECTRON_WDS_PORT}` : `file://${__dirname}/index.html`;
```

In case we agree on the change and merge the PR, I will create an additional PR to reflect this in https://github.com/electron-userland/electron-webpack-quick-start/blob/master/src/main/index.js#L18

**Comment by loopmode** [comment]
I found it very tricky to generate the port number in one place and transport it to where it is needed (dev-runner needs it for electronProcess env, RendererTarget needs it for webpack-dev-server)

The latest commit achieves it by bluntly setting `process.env.ELECTRON_WDS_PORT` in `main.ts`.
This is not pretty, as we are using a global context for value transport.
However, it works pretty well without adding too many changes.

**Comment by develar** [comment]
Merging in process, thanks.

I renamed `ELECTRON_WDS_PORT` to `ELECTRON_WEBPACK_WDS_PORT` because although we state fact that is an internal env, it is used in the `RendererTarget`, so, user can use it to set custom WDS port if provided `dev-runner` is not used.

**Comment by develar** [comment]
And on macOS you must also pass `host: "localhost"` since it possible to listen the same port in different interfaces (localhost vs 0.0.0.0). Will be fixed as part of merge.

**Comment by ThisNameNotUsed** [comment]
> Merging in process, thanks.
> 
> I renamed `ELECTRON_WDS_PORT` to `ELECTRON_WEBPACK_WDS_PORT` because although we state fact that is an internal env, it is used in the `RendererTarget`, so, user can use it to set custom WDS port if provided `dev-runner` is not used.

Wait. How? I have been trying to get the port number to pass to a script in src/common all day but, even after finding this GitHub thread, I can't figure out how to do it. I thought ELECTRON_WEBPACK_WDS_PORT would be exposed in the common folder but it is not. Is there any way you can make it so setting 

```
devServer: {
        contentBase: path.join(__dirname, "dist"),
        port: 3000 //<<<---------- HERE
    }

```
 in webpack.renderer.additions.js sets the port number like it does in WDS's normal usage or convert to us webpack noobs how to do it in the online documentation?

**Pull request closed without merging**